### PR TITLE
Performance tests.

### PR DESCRIPTION
**Pull request by dkorolev**
The actual commit: https://github.com/C5T/Current/commit/1de35cf3cc0a75a8228214c3f1eaa258db74faa1

(As this PR incorporates https://github.com/C5T/Current/pull/506 as of now).

Max, I'm seeing 33K QPS on Current's HTTP server on Hetz, and 16K..18K on my T420s. Fucking impressive.

At the same time, an attempt to test the QPS of a plain nginx-based `return 200 "OK"` fails for the reason I don't understand. Looks like nginx config update does go through, but sometimes fails to stick or something.

On the other hand, the issue may be with Current's HTTP client, somehow bringing the port into some unresponsive state. Here's what I can reproduce even for no-nginx test:

```
dima@dima-i9: ~/github/dkorolev/Current/examples/Benchmark/generic(perftest) $ ./.current/run --scenario=current_http_server
16633.6 QPS.
dima@dima-i9: ~/github/dkorolev/Current/examples/Benchmark/generic(perftest) $ ./.current/run --scenario=current_http_server
18117.6 QPS.
dima@dima-i9: ~/github/dkorolev/Current/examples/Benchmark/generic(perftest) $ ./.current/run --scenario=current_http_server
16392.8 QPS.
dima@dima-i9: ~/github/dkorolev/Current/examples/Benchmark/generic(perftest) $ ./.current/run --scenario=current_http_server
16229.2 QPS.
dima@dima-i9: ~/github/dkorolev/Current/examples/Benchmark/generic(perftest) $ ./.current/run --scenario=current_http_server
terminate called after throwing an instance of 'current::net::SocketConnectException'
  what():  ./../../../Sherlock/../Blocks/HTTP/../../Bricks/net/http/impl/../../tcp/impl/posix.h:560	SocketConnectException()
Aborted (core dumped)
dima@dima-i9: ~/github/dkorolev/Current/examples/Benchmark/generic(perftest) $ 
dima@dima-i9: ~/github/dkorolev/Current/examples/Benchmark/generic(perftest) $ ./.current/run --scenario=current_http_server
terminate called after throwing an instance of 'current::net::SocketConnectException'
  what():  ./../../../Sherlock/../Blocks/HTTP/../../Bricks/net/http/impl/../../tcp/impl/posix.h:560	SocketConnectException()
Aborted (core dumped)
dima@dima-i9: ~/github/dkorolev/Current/examples/Benchmark/generic(perftest) $ ./.current/run --scenario=current_http_server
terminate called after throwing an instance of 'current::net::SocketConnectException'
  what():  ./../../../Sherlock/../Blocks/HTTP/../../Bricks/net/http/impl/../../tcp/impl/posix.h:560	SocketConnectException()
Aborted (core dumped)
```

Could you try on your Mac please?

Thanks,
Dima

### CODE REVIEW

**Comment by dkorolev** [comment]
Just in case: when carefully run with nginx configured manually, Current's HTTP server is on par with nginx.

```
dima@dima-i9: ~/github/dkorolev/Current/examples/Benchmark/generic(perftest) $ git diff --no-ext-diff 
diff --git a/examples/Benchmark/generic/scenario_nginx_client.h b/examples/Benchmark/generic/scenario_nginx_client.h
index 2ceab7c..f3068cf 100644
--- a/examples/Benchmark/generic/scenario_nginx_client.h
+++ b/examples/Benchmark/generic/scenario_nginx_client.h
@@ -63,7 +63,7 @@ SCENARIO(simple_nginx_client, "Use Current's HTTP client to access an nginx-cont
     server_directive.CreateLocation(FLAGS_nginx_client_local_route)
         .Add(SimpleDirective("return", "200 \"" + FLAGS_nginx_client_test_body + "\""));
 
-    nginx_manager_.UpdateConfig(std::move(server_directive));
+    //nginx_manager_.UpdateConfig(std::move(server_directive));
   }
 
   void RunOneQuery() override { assert(HTTP(GET(url_)).body == FLAGS_nginx_client_test_body); }
dima@dima-i9: ~/github/dkorolev/Current/examples/Benchmark/generic(perftest) $ cat /etc/nginx/sites-enabled/current_perftest
server {
  listen 11111;
  location /perftest {
    return 200 "200 OK generated by nginx!";
  }
  location /ae00acdc38ead2003665d84d2dd22a7e9de58b955ac9509bea2f45643f83809f {
    return 200 "ae2fb999cd61669524c909fc12d9000646d348484ddeffc84550f451f099e446";
  }
}
dima@dima-i9: ~/github/dkorolev/Current/examples/Benchmark/generic(perftest) $ curl localhost:11111/perftest && echo
200 OK generated by nginx!
dima@dima-i9: ~/github/dkorolev/Current/examples/Benchmark/generic(perftest) $ ./.current/run --scenario=simple_nginx_client
17043.6 QPS.
dima@dima-i9: ~/github/dkorolev/Current/examples/Benchmark/generic(perftest) $ ./.current/run --scenario=current_http_server
17904 QPS.
```

Tested with less threads too. `nginx`'s QPS is higher on `--threads=1` and `--threads=4`, but it's already even with `threads` >= 20.

**Comment by dkorolev** [comment]
And confirmed the length of the message from a `Current`-powered HTTP server does not affect performance on my local machine.

**Comment by dkorolev** [comment]
@mzhurovich , https://github.com/C5T/Current/pull/507/commits/a244ded50f47df139d95a8e2af7305ac1c97c95c introduces port spanning.

Current's HTTP server is officially kicking nginx's ass ... on some machines, but still.

Thanks,
Dima

File: examples/Benchmark/generic/run.cc [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+          thread_(&Thread::ThreadFunction, this) {}
+
+    void Join() { thread_.join(); }
+
+    const SCENARIO& scenario_;
+    const double wall_time_second_end_;
+    size_t queries_completed_within_desired_timeframe_;
+    std::thread thread_;
+
+    // The thread function runs the queries continuously. It only count the queries
```
Comment (mzhurovich) [inline examples/Benchmark/generic/run.cc]: count**s**

File: examples/Benchmark/generic/run.cc [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    // completed witin the originally desired number of seconds in the final number.
```
Comment (mzhurovich) [inline examples/Benchmark/generic/run.cc]: s/witin/within/

File: examples/Benchmark/generic/scenario_golden_1k_qps.h [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+#ifndef CURRENT_MAKE_CHECK_MODE
+DEFINE_double(golden_qps, 1000, "Golden QPS for the `golden_1kqps` scenario to use.");
+#else
+DECLARE_double(golden_qps);
+#endif
+
+// clang-format off
+SCENARIO(golden_1kqps, "golden 1000 QPS") {
+  // Don't use `current::time::Now()`, as it's:
+  // a) under am unnecessary mutex, and
```
Comment (mzhurovich) [inline examples/Benchmark/generic/scenario_golden_1k_qps.h]: s/am/an/

File: examples/Benchmark/generic/scenario_golden_1k_qps.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  // b) guaranteed to increase by at least 1 per call.
+  static std::chrono::microseconds LocalNow() {
+    return std::chrono::duration_cast<std::chrono::microseconds>(
+        std::chrono::system_clock::now().time_since_epoch());
+  }
+
+  void RunOneQuery() override{
```
Comment (mzhurovich) [inline examples/Benchmark/generic/scenario_golden_1k_qps.h]: Space after `override`.

File: examples/Benchmark/generic/scenario_nginx_client.h [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+  simple_nginx_client() {
+    using namespace current::nginx::config;
+
+    for (uint16_t port = FLAGS_nginx_client_local_port_begin; port < FLAGS_nginx_client_local_port_end;
+         ++port) {
+      current::nginx::NginxManager nginx(FLAGS_nginx_client_config_file_prefix +
+                                         current::strings::ToString(port));
+
+      FullConfig config;
```
Comment (mzhurovich) [inline examples/Benchmark/generic/scenario_nginx_client.h]: Looks like `config` is not used anywhere.

File: examples/Benchmark/generic/run.cc (hunk 1 above)
Comment (dkorolev) [inline examples/Benchmark/generic/run.cc]: :+1: ea4b018

File: examples/Benchmark/generic/run.cc (hunk 2 above)
Comment (dkorolev) [inline examples/Benchmark/generic/run.cc]: :+1: ea4b018

File: examples/Benchmark/generic/scenario_golden_1k_qps.h (hunk 3 above)
Comment (dkorolev) [inline examples/Benchmark/generic/scenario_golden_1k_qps.h]: :+1: ea4b018

File: examples/Benchmark/generic/scenario_golden_1k_qps.h (hunk 4 above)
Comment (dkorolev) [inline examples/Benchmark/generic/scenario_golden_1k_qps.h]: :+1: ea4b018

**Comment by dkorolev** [comment]
@mzhurovich PTAL!

**Pull request merged**